error.go

```diff
@@ -27,39 +27,45 @@ var _ fmt.Formatter = (*Error)(nil)
 // Dynamic properties is a brittle mechanism and should therefore be used with care and in a simple and robust manner.
 func (e *Error) WithProperty(key Property, value interface{}) *Error {
 	errorCopy := *e
+	errorCopy.properties = errorCopy.properties.with(key, value)
+	return &errorCopy
+}
 
-	if errorCopy.properties == nil {
-		errorCopy.properties = make(map[Property]interface{}, 1)
-	} else {
-		errorCopy.properties = make(map[Property]interface{}, len(e.properties) + 1)
-		for k, v := range e.properties {
-			errorCopy.properties[k] = v
-		}
+// WithNewProperty adds a dynamic property to error instance if it were not already set.
+// If an error already contained another value for the same property, it is NOT overwritten.
+func (e *Error) WithNewProperty(key Property, value interface{}) *Error {
+	if _, ok := e.Property(key); ok {
+		return e
 	}
-
-	errorCopy.properties[key] = value
+	errorCopy := *e
+	errorCopy.properties = errorCopy.properties.with(key, value)
 	return &errorCopy
 }
 
 // WithUnderlyingErrors adds multiple additional related (hidden, suppressed) errors to be used exclusively in error output.
 // Note that these errors make no other effect whatsoever: their traits, types, properties etc. are lost on the observer.
 // Consider using errorx.DecorateMany instead.
 func (e *Error) WithUnderlyingErrors(errs ...error) *Error {
-	errorCopy := *e
-
-	newUnderying := make([]error, 0, len(e.underlying) + len(errs))
-	newUnderying = append(newUnderying, e.underlying...)
+	underlying := e.underlying()
+	newUnderlying := underlying
 
 	for _, err := range errs {
 		if err == nil {
 			continue
 		}
 
-		newUnderying = append(newUnderying, err)
+		newUnderlying = append(newUnderlying, err)
 	}
 
-	errorCopy.underlying = newUnderying
-	return &errorCopy
+	if len(newUnderlying) == len(underlying) {
+		return e
+	}
+
+	l := len(newUnderlying)
+	newUnderlying = newUnderlying[:l:l]
+	errorCopy := e.WithProperty(propertyUnderlying, newUnderlying)
+	errorCopy.hasUnderlying = true
```

Comment:
```suggestion
	errorCopy.hasUnderlying = l > 0
```

Reply:
No, because length could only grow, and not grown length == 0 were caught couple of lines above.

Reply:
True. Still, it may not be cleat at the first glance, and some future change may accidentally break this, but that's a minor point. 
Ideally, the code should be as I suggested, and the compiler should eliminate the 'always true' calculation.